File: code/q1_1.m
%% 习题1 第17题
% 课本 P20
clc
clear

prompt = '请输入 N: ';
N = input(prompt);

data0 = SN_Real(N);
data1 = SN_1(N);
data2 = SN_2(N);
fprintf('准确值:\t%.10f\n', data0);
fprintf('正向求和:\t%.10f, 误差: %.10f\n', data1, abs(data0 - data1));
fprintf('反向求和:\t%.10f, 误差: %.10f\n', data2, abs(data0 - data2));

%% 画图
clc
clear

prompt = '请输入 N: ';
N = input(prompt);
N_start = 101;

x = N_start:N;
y1 = zeros(1, (N-N_start+1));
y2 = zeros(1, (N-N_start+1));

tic
for i = x
    data0 = SN_Real(i);
    data1 = SN_1(i);
    data2 = SN_2(i);
%     y1(i - N_start + 1) = abs(data0 - data1);
%     y2(i - N_start + 1) = abs(data0 - data2);
    y1(i - N_start + 1) = (data0 - data1);
    y2(i - N_start + 1) = (data0 - data2);
end
toc

figure
plot(x, y1, x, y2);
legend('从大到小', '从小到大');
title('误差（准确值-计算值）')
xlabel('N');
ylabel('e');

saveas(gcf,'e.jpg')%保存图像为文件

%% ----------------------
function res = f(N)
%f - 1/(N^2 - 1)
%
% Syntax: res = f(N)
%
% 
    res = single(1/(N^2 - 1));
end

function sum = SN_1(N)
%SN_1 - sum f(a), a from 2 to N
%
% Syntax: res = SN_1(N)
%
% 正向求和，从 2 到 N
    sum = 0;
    for i = 2:N
        sum = sum + f(i);
    end
    sum = single(sum);
end

function sum = SN_2(N)
%SN_2 - sum f(a), a from N to 2
%
% Syntax: sum = SN_2(N)
%
% 反向求和，从 N 到 2
    sum = 0;
    for i = 2:N
        sum = sum + f(N-i+2);
    end
    sum = single(sum);
end

function sum = SN_Real(N)
%准确值
    sum = single((3/2 - (1/N) - (1/(N-1)))/2);
end

File: code/chapter1/q1-1.cpp
#include <iostream>
#include <iomanip>
#include <math.h>

float f(int N) {
    float res = 0;
    res = float(1.0)/(pow(float(N), float(2)) - 1);
    return res;
}

float SN_1(int N) {
    float sum = 0;
    for (int i = 2; i <= N; i++)
    {
        sum += f(i);
    }
    return sum;
}

float SN_2(int N) {
    float sum = 0;
    for (int i = N; i >= 2; i--)
    {
        sum += f(i);
    }
    return sum;
}

float SN_Real(int N) {
    double sum = 0.0;
    // sum = float(0.5)*(float(1.5) - float(1.0)/float(N) - float(1.0)/float(N+1.0));
    sum = 0.5*(1.5 - 1.0/double(N) - 1.0/double(N+1.0));
    return sum;
}

int main() {
    float data0 = 0.0;
    float data1 = 0.0;
    float data2 = 0.0;

    int N = 0;

    std::cout<<"请输入N:"<<std::endl;
    std::cin>>N;

    data0 = SN_Real(N);
    data1 = SN_1(N);
    data2 = SN_2(N);

    std::cout<<"N\t精确值\t\t从大到小\t误差1   \t从小到大\t误差2"<<std::endl;
    std::cout<<N<<"\t"<< std::fixed << std::setprecision(16)<<data0<<"\t"<<data1<<"\t"<<abs(data0-data1)<<"\t"<<data2<<"\t"<<abs(data0-data2)<<std::endl;

    // printf("          value                              error\r\n");
    // printf("准确值:   %.24f\r\n", data0);
    // printf("从大到小: %.24f, 误差: %.24f\r\n", data1, abs(data0-data1));
    // printf("从大到小: %.24f, 误差: %.24f\r\n", data2, abs(data0-data2));

    return 0;
}